### PR TITLE
Change the logic flow for the order confirmation thankyou screen

### PR DESCRIPTION
**Pull request by coreymckrill**
### Changes proposed in this Pull Request:

On the heels of #38983 and #39191, this changes the flow a bit so that the "thank you" message will always be shown when loading the "order received" URL. However, the verification requirement still prevents the details of the order being shown unless the customer is verified, either by confirming the email or by logging in. If unverified, the login/email form is shown on the thank you screen instead of the details of the order.

This attempts to solve cases where a newly created order has a customer record attached to it (i.e. is not a guest checkout) but the customer is not actually logged in. Currently this can cause the customer to see a login screen right after completing checkout, rather than an acknowledgment that the order has been received. This ensures that the customer gets that acknowledgment. Further, it allows developers to modify this part of their custom checkout flows, if necessary, by customizing the thankyou template file.

See [this conversation](https://github.com/woocommerce/woocommerce/pull/38983#issuecomment-1633890821) for more details on the issues this is attempting to resolve.

### To do

- [ ] If this seems like a worthwhile change to pursue, the e2e tests will probably need to be updated.

### How to test the changes in this Pull Request:

The instructions in #39191 apply to this PR as well. The only change should be that the "Thank you. Your order has been received." message should be shown even when viewed as an unverified customer.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

Change the order confirmation screen to always show the 'Thank you for your order' message, even if the viewer must log in or confirm the order email to see the details of the order.

#### Comment

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php [hunk 1]
```diff
@@ -297,29 +297,19 @@ private static function order_received( $order_id = 0 ) {
 			return;
 		}
 
+		$template_args     = array( 'order' => $order );
```
Comment (barryhughes) [inline plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php]: (1/3) Pretty minor, but I wonder if this should be:

```suggestion
		$template_args     = array( 
			'needs_verification' => false,
			'order'              => $order,
		);
```

File: plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php [hunk 2]
```diff
@@ -297,29 +297,19 @@ private static function order_received( $order_id = 0 ) {
...
-					'verify_url'        => $order->get_checkout_order_received_url(),
-				)
-			);
-			return;
+		if ( ! $order_customer_id && self::guest_should_verify_email( $order, 'order-received' ) ) {
+			// For guest orders, request they verify their email address (unless we can identify them via the active user session).
+			$template_args['needs_verification'] = 'email';
+		} elseif ( $order_customer_id && get_current_user_id() !== $order_customer_id ) {
+			// For non-guest orders, require the user to be logged in before showing this page.
+			$template_args['needs_verification'] = 'login';
```
Comment (barryhughes) [inline plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php:311]: (2/3) On the basis that these conditions may not be satisfied and therefore `$template_args['needs_verification']` may not be set.

File: plugins/woocommerce/templates/checkout/thankyou.php [hunk 3]
```diff
@@ -41,43 +41,70 @@
 
 			<p class="woocommerce-notice woocommerce-notice--success woocommerce-thankyou-order-received"><?php echo apply_filters( 'woocommerce_thankyou_order_received_text', esc_html__( 'Thank you. Your order has been received.', 'woocommerce' ), $order ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?></p>
 
-			<ul class="woocommerce-order-overview woocommerce-thankyou-order-details order_details">
+			<?php if ( $needs_verification ) : ?>
```
Comment (barryhughes) [inline plugins/woocommerce/templates/checkout/thankyou.php:44]: (3/3) However, we still test it here (and, depending on the version of PHP in use, I think may result in a notice level error or higher being emitted).

**Review by barryhughes — COMMENTED** [review]
This seems like a nice tweak, since we always acknowledge the order was received. Left a suggestion, otherwise I think this is worth moving ahead with.

**Comment by coreymckrill** [comment]
@barryhughes Ok, I committed your suggestion and fixed up some phpcs issues. This is ready for another pass.

**Review by barryhughes — COMMENTED** [review]
Something I think we should pause and consider (apologies for not raising this during my first pass) is that, in 7.9.0, if we decided extra verification was needed before showing the order details we would either:

- [Render the `form-verify-email` template](https://github.com/woocommerce/woocommerce/blob/7.9.0/plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php#L302-L307) (for non-guest orders)
- *Or,* [render a notice plus the login form](https://github.com/woocommerce/woocommerce/blob/7.9.0/plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php#L309-L319) (for guest orders)

We then returned early and did not render any of the `thankyou` template. Whereas, in this change, we always render at least some of the `thankyou` template (and of course it may return early after embedding either of the above). 

I think this may work against the motivation for adding verification checks because, if the order exists but the shopper is unverified, we are still going to trigger the `woocommerce_before_thankyou` action via code found close to the start of the `thankyou` template (and a third party extension or similar may unwittingly expose information via that hook that perhaps should not be exposed to the as-yet unverified visitor).

There's also some asymmetry in these scenarios where verification is needed in the sense that we do still trigger `woocommerce_before_thankyou` but not `woocommerce_thankyou_<METHOD>` or `woocommerce_thankyou`. However, this is less of a concern than the previous point.

What do you think of an alternative in which we essentially roll back to the previous structure, but add the *'your order has been received'* messaging before rendering either the email verification or login forms? 

Also, though far less pressing/less essential: the *'your order has been received'* verbiage and its accompanying hook is already duplicated twice, so that could possibly also be encapsulated into a function of its own.

**Comment by barryhughes** [comment]
...yet another alternative could be restructuring the template to something like:

```
if ( order ) {
    # PRINT THANK YOU/ORDER RECEIVED

    if ( needs_verification ) {
        # VERIFICATION
    } else {
        # SHOW ORDER DETAILS
    }
} else {
    # FINAL FALLBACK
}
```

With the *possible* disadvantage that sites with existing template overrides will not gain this additional layer of protection.

**Comment by coreymckrill** [comment]
@barryhughes 

> What do you think of an alternative in which we essentially roll back to the previous structure, but add the 'your order has been received' messaging before rendering either the email verification or login forms?
>
> Also, though far less pressing/less essential: the 'your order has been received' verbiage and its accompanying hook is already duplicated twice, so that could possibly also be encapsulated into a function of its own.

This sounded worth trying, so I've opened #39758 as an alternative to this, and added you as the reviewer.

**Comment by barryhughes** [comment]
- Resolved via https://github.com/woocommerce/woocommerce/pull/39758.
- Closing this one.

**Pull request closed without merging**